ThorsSerializer 1.5.7 (new formula)

Loki-Astari opened the pull request:
Modern Declarative Serializer for C++

    Allows developers to declare what classes can be serialized.
    The serialization library will then serialize object of that type to Json/Yaml.

- [ ] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?

Yes

- [ ] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?

Yes

- [ ] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?

Yes

- [ ] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

Yes

-----

SMillerDev commented:
Your submodule fix doesn't seem to work. Is there no way you could add full source packages to your releases?

In addition to that I don't believe it would pass brew audit. There's no version information and no test, both of which should be tested by the audit.

Loki-Astari commented:
> Your submodule fix doesn't seem to work. Is there no way you could add full source packages to your releases?

I will look into the build issue here.
I am pretty sure I can fix these (I just had not tried to build like this before).
Is there a list of ways to build that I can try independently?

> In addition to that I don't believe it would pass brew audit.
> There's no version information

The code is tagged with 1.5.7 (I though that was the version information you required).
Is there something different I should be doing.

> and no test, both of which should be tested by the audit.
This library comes with extensive tests. These are done as part of the install. Any install must pass all the tests. But they can be done separately with `make test`.

If I remove the:
````
  test do
    system "false"
  end
````
Would that satisfy the requirements.

SMillerDev commented:
All formula need a test block. This block should test if the program is actually functioning after installing it.

Loki-Astari commented:
After looking threw the other formula I managed to come up with this.

```` 
  test do
    (testpath/"test.cpp").write <<~EOS
        #include "ThorSerialize/JsonThor.h"
        #include "ThorSerialize/SerUtil.h"
        #include <sstream>
        #include <iostream>
        #include <string>
        #include <vector>

        struct Block
        {
            std::string             key;
            int                     code;
        };
        ThorsAnvil_MakeTrait(Error, key, code);

        int main()
        {
            using ThorsAnvil::Serialize::jsonImport;
            using ThorsAnvil::Serialize::jsonExport;

            std::stringstream   inputData(R"({"key":"XYZ","code":37373})");

            Block    object;
            inputData >> jsonImport(object);

            if (object.key != "XYZ" || object.code != 37373) {
                std::cerr << "Fail\n";
                return 1;
            }
            std::cerr << "OK\n";
            return 0;
        }
    EOS
    system ENV.cc, "test.cpp", "-o", "test",
                   "-I#{include}",
                   "-L#{lib}",
                   "-llibThorSerialize17"
    system "./test"
  end
````

I still have to test and stuff but is this on the correct path of what you are looking for?

SMillerDev commented:
Obviously I'm not very familiar with the tool but the structure looks good.

Loki-Astari commented:
Just waiting for a clone a clean virtual mac to do the finish so I can test it.
Then ill push the new commit.

SMillerDev commented:
I think you need to read "Specifying the download strategy explicitly
" in the [Formula Cookbook](https://docs.brew.sh/Formula-Cookbook). Cause the download is currently failing.

Loki-Astari commented:
I explicitly took that out (specifying the download stratergy) because when i ran
````
    brew audit ThorsSerializer --online --new-formula
````
It was saying the :using value was superfluous.

I'll add it back and try again.

Loki-Astari commented:
Well at least I will have a blog article out of this.
I now know about 4 commands that you should run locally and make sure work before submitting. :-)

DomT4 commented on Formula/ThorsSerializer.rb:
```diff
@@ -0,0 +1,57 @@
+  desc "Declarative Serialization Library (Json/Yaml) for C++"
+  homepage "https://github.com/Loki-Astari/ThorsSerializer"
+  url "https://github.com/Loki-Astari/ThorsSerializer.git",
+    :using => :git,
```
This is unnecessary.

Loki-Astari replied:
Tested. Correct. Removed.

DomT4 commented on Formula/ThorsSerializer.rb:
```diff
@@ -0,0 +1,57 @@
+    :tag => "1.5.7",
+    :revision => "c391b49a36c7432a3fadbbb3e602bba1f61d4078"
+
+  ENV["COV"] = "gcov"
```
This is in the wrong place. I presume you want it inside `def install`.

Loki-Astari replied:
Yep that seems like a good idea.

DomT4 commented on Formula/ThorsSerializer.rb:
```diff
@@ -0,0 +1,57 @@
+  depends_on "libyaml"
+
+  def install
+    system "./configure", "--disable-binary", "--disable-vera", "--with-thor-build-on-travis", "--prefix=#{prefix}"
```
Wrap to 80 characters.

Loki-Astari replied:
No problem.

Loki-Astari commented:
@SMillerDev , Morning; Is there anything else I should be doing?

Not sure of the speed of the processes.
Any idea on the time frame this type of thing takes?

SMillerDev commented:
It differs a lot between formula, it depends a lot on peoples time to review changes. Maybe @DomT4 has time to look at it again though?

DomT4 commented:
FWIW I can review stuff _(and am happy to do so!)_ but I stepped away from being able to merge stuff a fair while ago now, so the final review/say/merge decision will always come down to @ilovezfs, @commitay or anyone else maintaining core these days.

DomT4 commented on Formula/ThorsSerializer.rb:
```diff
@@ -0,0 +1,59 @@
+
+    system "./configure", "--disable-binary",
+                          "--disable-vera",
+                          "--with-thor-build-on-travis",
```
Out of wild curiosity, what is this argument doing/impacting?

Loki-Astari replied:
--disable-binary:   There is an experimental binary serializer/deserialization in this lib.
                                I want to convert this to use a "standard" binary serializer;
                                something like googles protocol buffer.

                                It's not needed for the primary use case so disabled here.

--disable-vera:       Vera++ does code style checking.
                                If people want to contribute code back they need to make sure it
                                adheres to the style guide lines. But it requires a whole bunch of
                                other libraries to work. It is not needed to simply build the library
                                so disable it to remove dependencies required.

--with-thor-build-on-travis:   I included all my submodules via the `git@github.com/XXX`
                                                The travis-ci.org does not support this. So I added this flag
                                                to support automated builds on travis. Its useful here becuase
                                                neither does your build system. Internally it just converts the
                                                submodules to use `https://github.com/XXX`

ilovezfs replied:
I'm confused how that works for submodules given we're long passed the point at which they're fetched by the time we hit configure

Loki-Astari replied:
In my initial pull request I had overridden the `update_submodules` in `GitHubGitDownloadStrategy` to try and use `git@github.com/XXX` and it worked for the normal build. 

````
class GitWithGitAt < GitHubGitDownloadStrategy
  def update_submodules
````

But it failed when building the binaries.  

It would be nice if you supported `git@github.com/XXX` submodules as it makes working with submodules easier (I don't need to check them out separately to update them). You would need to create a SSH deployment key that people could add to their github repositories.

ilovezfs commented on Formula/ThorsSerializer.rb:
```diff
@@ -0,0 +1,59 @@
+  desc "Declarative Serialization Library (Json/Yaml) for C++"
+  homepage "https://github.com/Loki-Astari/ThorsSerializer"
+  url "https://github.com/Loki-Astari/ThorsSerializer.git",
+    :tag => "1.5.7",
```
: under "

Loki-Astari replied:
Sorry. I am not sure what you mean.

ilovezfs replied:
more spaces until the : in :tag is under " in "https://github.com/Loki-Astari/ThorsSerializer.git",

DomT4 replied:
> : under "

Didn't know I'd managed to win this indentation battle! 🎉

Loki-Astari replied:
No problem.

ilovezfs replied:
@DomT4 until the Ruby police decide it should be a two-space continuation because they said so.

DomT4 replied:
@ilovezfs I'll go along with a two-space indent on most things but aligning URLs like that bothers me more than it should, heh.

ilovezfs commented on Formula/ThorsSerializer.rb:
```diff
@@ -0,0 +1,59 @@
+class Thorsserializer < Formula
+  desc "Declarative Serialization Library (Json/Yaml) for C++"
```
why is this title case

Loki-Astari replied:
No reason. Will fix.

Loki-Astari commented:
I just remembered. 
I could not get the submodule stuff working in your environment.

So I converted them all to `https://github.com/XXX` and removed that flag.
I'll remove it from the Formula now.

Loki-Astari commented on Formula/ThorsSerializer.rb:
```diff
@@ -0,0 +1,59 @@
+  depends_on "libyaml"
+
+  def install
+    ENV["COV"] = "gcov"
```
In your build environment the configuration script fails to detect the compiler that will be used and thus uses the wrong tool to do code coverage analysis of the unit tests (this causes the build to fail).

By setting this environment variable we override the configuration script and specify the exact code coverage tool to use. This allows the code coverage to correctly be done during the build.

Loki-Astari commented:
@ilovezfs I will change the formula file name.

DomT4 commented:
> camel case formula file name looks fishy to me

Amused I didn't spot this 😅.

ilovezfs commented:
do we prefer
```
brew install thorsserializer
```
or
```
brew install thors-serializer
```

Loki-Astari commented:
@ilovezfs If that's a question for me; then

I like `thors-serializer`
I have a couple of other libs under the `thors-` label that are useful.

ilovezfs commented:
then file name thors-serializer.rb and class name ThorsSerializer

ilovezfs commented:
I see C++14 in the test. Should formula have a `needs :cxx14`?

Loki-Astari commented:
@ilovezfs The configuration script run a test on the compiler to determine what flags are needed to run the compiler in the appropriate mode. If the compiler can not support C++14 then the configuration script would error out.

So does it need it in the Formula if the configuration script is already doing the check?

Loki-Astari commented:
I think its safe.  
It is 2018 after all.    
Most (I bet all) compiler support (with flags) a 4 year old version of the standard. 😄

SMillerDev commented:
There's also people using homebrew on older systems so it can't hurt to define it iyam

Loki-Astari commented:
No problem. I will do a quick google to find where to add "needs :cxx14"
Let me test it on my virtual box then I will push a new version.

ilovezfs commented on Formula/thors-serializer.rb:
```diff
@@ -0,0 +1,59 @@
+      :tag => "1.5.7",
+      :revision => "c391b49a36c7432a3fadbbb3e602bba1f61d4078"
+
+  depends_on "libyaml"
```
blank line under here

ilovezfs commented on Formula/thors-serializer.rb:
```diff
@@ -0,0 +1,60 @@
+class ThorsSerializer < Formula
+  desc "Declarative serialization library (Json/Yaml) for C++"
```
JSON and YAML are all caps

Loki-Astari replied:
Will do.

But a personal comment: It looks uglier.

ilovezfs replied:
Maybe one day they will be spelled Jason and Yamel.

Loki-Astari replied:
💯 😈

DomT4 replied:
> spelled

Americans, sigh.

Loki-Astari commented:
@ilovezfs All building and green again 😄 

Let me know if there is anything else I can do.  
Thanks for all the pointers.

I think I may have enough material to write a blog article on how to go about writing a homebrew "Formula". 🚀 💥

DomT4 commented:
The [Cookbook](https://docs.brew.sh/Formula-Cookbook) is practically the size of a novella if you need further inspiration 😆.

This PR is still good with me, but you'll need to await Joe's final glance over everything.
